onnx/defs/nn/defs.cc

```diff
@@ -2074,4 +2074,41 @@ ONNX_OPERATOR_SET_SCHEMA(
              {{"Processed_STD"}, "Add", {"STD", "Epsilon"}},
              {{"Y"}, "Div", {"X_variance", "Processed_STD"}}})));
 
+static const char* celu_ver12_doc = R"DOC(
+       Continuously Differentiable Exponential Linear Units:
+       Perform the linear unit element-wise on the input tensor X
+       using formula: <br/> ``` max(0,x) + min(0,α*(exp(x/α)−1)) ```
+)DOC";
+
+static float celu_default_alpha = 1.0;
+
+ONNX_OPERATOR_SET_SCHEMA(
+    Celu,
+    12,
+    OpSchema()
+       .SetDoc(celu_ver12_doc)
+       .Input(0, "X", "Input tensor", "T")
+       .Output(0, "Y", "Output tensor", "T")
+       .Attr(
+           "alpha",
+           "The Alpha value in Celu formula which control the shape of "
+           "the unit. The default value is 1.0.",
+           AttributeProto::FLOAT,
+           celu_default_alpha)
+       .TypeConstraint(
+           "T",
+           {"tensor(float16)", "tensor(float)", "tensor(double)"},
+           "Constrain input and output types to floating-point tensors.")
+       .FunctionBody(FunctionBodyHelper::BuildNodes(
+           {// nodes: {outputs, op, inputs, attributes}
+            //FunctionBodyHelper::NodeDef{{"alpha"}, "Constant", {}, {{"value", 1.f}}},
+	    //FunctionBodyHelper::Const<float>("alpha", 1.0f),
+            //FunctionBodyHelper::Const<float>("Y", 1.0f),
+	    FunctionBodyHelper::NodeDef{{"Y"}, "Constant", {}, {MakeRefAttribute("value", AttributeProto::FLOAT, "alpha")}},
+            {{"X_alpha"},
+             "Div",
+             {"X", "alpha"}
+            },
+            {{"U"}, "Elu", {"X_alpha"}}})));
```

Comment:
From Pytorch, CELU equation is
```
CELU(x)=max(0,x)+min(0,α∗(exp(x/α)−1))
```
while ELU uses
```
ELU(x)=max(0,x)+min(0,α∗(exp(x)−1))
```
In addition, here the function body is doing
```
ELU(x/α)=max(0,x/α)+min(0,(exp(x/α)−1))
```
which doesn't exactly match Pytorch CELU. Is this expected? Or I miss something?

Reply:
😱

You are completely right, the ELU implementation of Pytorch is different from ONNX Elu, and it is not possible to express CELU from ONNX's ELU. Thanks you for noticing it, I am working on a fix.

Reply:
Here is a code testing the differences between Pytorch CELU and the implementation (with corrected parenthesis) I provided.

```
import numpy as np
import torch

def onnx_celu(input_data, alpha=1.0):
    positive_input = np.maximum(0, input_data)
    negative_input = np.minimum(0, alpha * (np.exp(input_data / alpha) - 1))
    output_data = positive_input + negative_input
    return output_data

def torch_celu(input_data, alpha=1.0):
    return torch.nn.CELU(alpha=alpha)(torch.Tensor(input_data)).numpy()

input_data = np.random.randn(1, 2, 3).astype('float32')
alpha = 2

assert (onnx_celu(input_data, alpha) == torch_celu(input_data, alpha)).all()

```